crates/subspace-archiving/src/archiver.rs

```diff
@@ -614,56 +631,82 @@ impl Archiver {
         Some(segment)
     }
 
-    /// Take segment as an input, apply necessary transformations and produce archived object mappings.
-    /// Must be called before `produce_archived_segment()`.
-    fn produce_object_mappings(&self, segment: &Segment) -> Vec<GlobalObject> {
+    /// Take the last block continuation in the next segment's items, apply necessary
+    /// transformations, and produce object mappings. Then remove the mappings from that block
+    /// continuation.
+    ///
+    /// Must be called on an incomplete segment.
+    fn produce_next_segment_mappings(&mut self) -> Vec<GlobalObject> {
+        let object_mapping = self.produce_object_mappings(self.buffer.iter());
+
+        // We've already produced these mappings, so we don't need to store them in the buffer.
+        if let Some(SegmentItem::BlockContinuation {
+            bytes: _,
+            object_mapping,
+        }) = self.buffer.back_mut()
+        {
+            object_mapping.objects_mut().clear();
+        }
+
+        object_mapping
+    }
+
+    /// Take the last block in the segment items, apply necessary transformations, and produce
+    /// object mappings.
+    ///
+    /// Must be called in the correct order with `produce_archived_segment()`, because that method
+    /// updates the archiver's state.
```

Comment:
This is an unnecessary comment, it is physically impossible to call in the incorrect order because `produce_archived_segment()` consumes the segment, so you'll not be able to call this method afterwards.

Reply:
The current wording is outdated. But this method can be called on `segment.items()` or `self.buffer.items()`, so it is possible to call it the wrong way.
 
I think this comment will be clearer:
```suggestion
    ///
    /// This method can be called on a segment’s items, or on the internal buffer. It must only be
    /// called on the buffer when all segments for a block have been produced. Before that, the
    /// buffer can contain a `BlockContinuation` which spans multiple segments.
```

Reply:
I would add a comment why it is okay to call it with a buffer right before call with a buffer and remove comment from the function itself. This function as such doesn't really know what it is called with and shouldn't care about it.